src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -80,24 +70,35 @@ internal enum PreconditionState
             }
 
             var serveBody = !HttpMethods.IsHead(request.Method);
-            if (HttpMethods.IsHead(request.Method) || HttpMethods.IsGet(request.Method))
+            var preconditionState = GetPreconditionState(context, httpRequestHeaders, lastModified, etag);
+            if (preconditionState == PreconditionState.NotModified)
             {
-                var preconditionState = GetPreconditionState(context, httpRequestHeaders, lastModified, etag);
-                if (request.Headers.ContainsKey(HeaderNames.Range) &&
-                    (preconditionState == PreconditionState.Unspecified ||
-                    preconditionState == PreconditionState.ShouldProcess))
-                {
-                    return SetRangeHeaders(context, httpRequestHeaders, fileLength, lastModified, etag);
-                }
-                if (preconditionState == PreconditionState.NotModified)
-                {
-                    serveBody = false;
-                    response.StatusCode = StatusCodes.Status304NotModified;
-                }
-                else if (preconditionState == PreconditionState.PreconditionFailed)
+                serveBody = false;
+                response.StatusCode = StatusCodes.Status304NotModified;
+            }
+            else if (preconditionState == PreconditionState.PreconditionFailed)
+            {
+                serveBody = false;
+                response.StatusCode = StatusCodes.Status412PreconditionFailed;
```

Comment:
Returning here would mean that this is never reached https://github.com/aspnet/Mvc/pull/6348/files#diff-bb6a01f81a4cd717b9350677639fd1f5R87